modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/insertbutton/plugin.js

```diff
@@ -220,98 +222,53 @@
 
 			const type = selection.getType();
 
-			if (type === CKEDITOR.SELECTION_ELEMENT || selection.isHidden()) {
-				this.hide();
-
-				return;
-			}
+			const startElement = selection.getStartElement();
 
 			if (
 				type === CKEDITOR.SELECTION_TEXT &&
-				selection.getSelectedText()
+				selection.getSelectedText() === '' &&
+				startElement.getText() === '\n'
 			) {
-				this.hide();
-
-				return;
-			}
+				this._positionButton();
 
-			const ranges = selection.getRanges();
-
-			const rangesNotCollapsed = CKEDITOR.tools.array.some(
-				ranges,
-				(range) => {
-					return !range.collapsed;
-				}
-			);
-
-			if (rangesNotCollapsed) {
-				this.hide();
-
-				return;
+				this.show();
 			}
-
-			const range = ranges[ranges.length - 1];
-
-			const startContainer = range.startContainer;
-
-			if (
-				typeof startContainer.getName === 'function' &&
-				startContainer.getName() === 'td'
-			) {
+			else {
 				this.hide();
-
-				return;
 			}
+		},
 
-			const nodeType = startContainer.$.nodeType;
-
-			if (nodeType === CKEDITOR.NODE_ELEMENT) {
-				if (startContainer.getChildCount() !== 1) {
-					this.hide();
-
-					return;
-				}
-				else if (startContainer.getChild(0)) {
-					const child = startContainer.getChild(0);
+		_positionButton() {
+			const selection = this.editor.getSelection();
 
-					if (
-						typeof child.getName === 'function' &&
-						child.getName() !== 'br'
-					) {
-						this.hide();
+			const startElement = selection.getStartElement();
 
-						return;
-					}
-				}
-			}
-			else if (
-				nodeType === CKEDITOR.NODE_TEXT &&
-				!startContainer.isEmpty()
-			) {
-				this.hide();
+			const selectionClientRect = startElement.getClientRect();
 
-				return;
-			}
+			const button = this._button;
 
-			const rangeClientRects = range.getClientRects(true);
+			const buttonStyles = window.getComputedStyle(button.$);
 
-			const rangeClientRect =
-				rangeClientRects[rangeClientRects.length - 1];
+			const buttonHeight = parseInt(buttonStyles.height, 10);
 
-			this._positionButton(rangeClientRect);
+			let sideOffset = selectionClientRect.x + BUTTON_SIDE_OFFSET;
 
-			this.show();
-		},
+			if (this.editor.config.contentsLangDirection === 'rtl') {
+				const buttonWidth = parseInt(buttonStyles.width, 10);
 
-		_positionButton(rangeClientRect) {
-			if (!this._buttonClientRect) {
-				this._buttonClientRect = this._button.getClientRect(true);
+				sideOffset =
+					selectionClientRect.x +
+					selectionClientRect.width -
+					buttonWidth -
+					BUTTON_SIDE_OFFSET;
 			}
 
-			this._button.setStyles({
-				left:
-					rangeClientRect.x - this._buttonClientRect.width / 2 + 'px',
-				top: rangeClientRect.y + 'px',
+			button.setStyles({
```

Comment:
This calculus looks nice. However it would be good use a library like `domAlign` for preventing these complex codes. Is there a limitation to use this library in this file? just asking because I never tested it and I don't remember clear thoughts to not use it like ES6.

Reply:
I don't think that the positioning problem in this case can be fixed by any static solution, CSS or a convenience library like `domAlign`. That's because CSS deals with _node_ styles, while our positioning is relative to selection within a text node. This requires [JS Selection API](https://developer.mozilla.org/en-US/docs/Web/API/Selection), wrapped by ckeditor.